core/src/main/java/hudson/security/SecurityRealm.java

```diff
@@ -290,15 +290,39 @@ public void doLogout(StaplerRequest req, StaplerResponse rsp) throws IOException
         Authentication auth = SecurityContextHolder.getContext().getAuthentication();
         SecurityContextHolder.clearContext();
 
-        // reset remember-me cookie
-        Cookie cookie = new Cookie(ACEGI_SECURITY_HASHED_REMEMBER_ME_COOKIE_KEY,"");
+        String contextPath = req.getContextPath().length() > 0 ? req.getContextPath() : "/";
+        resetRememberMeCookie(req, rsp, contextPath);
+        clearStaleSessionCookies(req, rsp, contextPath);
+
+        rsp.sendRedirect2(getPostLogOutUrl(req,auth));
+    }
+
+    private void resetRememberMeCookie(StaplerRequest req, StaplerResponse rsp, String contextPath) {
+        Cookie cookie = new Cookie(ACEGI_SECURITY_HASHED_REMEMBER_ME_COOKIE_KEY, "");
         cookie.setMaxAge(0);
         cookie.setSecure(req.isSecure());
         cookie.setHttpOnly(true);
-        cookie.setPath(req.getContextPath().length()>0 ? req.getContextPath() : "/");
+        cookie.setPath(contextPath);
         rsp.addCookie(cookie);
+    }
 
-        rsp.sendRedirect2(getPostLogOutUrl(req,auth));
+    private void clearStaleSessionCookies(StaplerRequest req, StaplerResponse rsp, String contextPath) {
+        for (Cookie cookie : req.getCookies()) {
+            if (!cookie.getName().startsWith("JSESSIONID."))
+                continue;
+            String cookieContextPath = cookie.getPath();
+            if (cookieContextPath == null) {
+                if (!contextPath.equals("/"))
+                    continue;
+            } else if (!cookieContextPath.equals(contextPath)) {
+                continue;
+            }
+            // one reason users log out is to clear their session(s)
+            // so tell the browser to drop all old sessions
+            cookie.setMaxAge(0);
+            cookie.setValue("");
+            rsp.addCookie(cookie);
+        }
```

Comment:
1) From my PoV the `continue` keyword should be avoided to prevent troubles.
2) The `if` should use `{}`
3) Consistency is really important to help readability of code

Proposal:
```
for (Cookie cookie : req.getCookies()) {
    if (cookie.getName().startsWith("JSESSIONID.")) {
        String cookieContextPath = cookie.getPath();
        boolean pathMatches;
        if (cookieContextPath == null) {
            pathMatches = contextPath.equals("/");
        } else {
            pathMatches = cookieContextPath.equals(contextPath);
        }
        if (pathMatches) {
            // one reason users log out is to clear their session(s)
            // so tell the browser to drop all old sessions
            cookie.setMaxAge(0);
            cookie.setValue("");
            rsp.addCookie(cookie);
        }
    }
}
```

Reply:
app servers and multiple paths in 2019 seem fairly edge cases (certainly the former) for the pain caused by this. If the worst case is a user session log out, not that bad?